src/ejabberd_receiver.erl

```diff
@@ -136,15 +140,24 @@ init([Socket, SockMod, Shaper, MaxStanzaSize]) ->
 %%                                      {stop, Reason, State}
 %% Description: Handling call messages
 %%--------------------------------------------------------------------
-handle_call({starttls, TLSSocket}, _From, State) ->
-    StateAfterReset = reset_parser(State),
-    NewState = StateAfterReset#state{socket = TLSSocket,
-                                 sock_mod = fast_tls},
-    case fast_tls:recv_data(TLSSocket, <<"">>) of
-        {ok, TLSData} ->
-            {reply, ok, process_data(TLSData, NewState), ?HIBERNATE_TIMEOUT};
-        {error, _Reason} ->
-            {stop, normal, ok, NewState}
+handle_call(get_socket, _From, #state{socket = Socket} = State) ->
+    {reply, {ok, Socket}, State, ?HIBERNATE_TIMEOUT};
+handle_call({starttls, TLSOpts}, From, #state{parser = Parser,
+                                              socket = TCPSocket} = State) ->
+    deactivate_socket(State),
+    gen_server:reply(From,ok), %% let ejabberd_socket finish tcp negotiations
+    case ejabberd_tls:tcp_to_tls(TCPSocket,TLSOpts) of
+        {ok,TLSSocket} ->
+            StateAfterReset = reset_parser(State),
+            NewState = StateAfterReset#state{socket = TLSSocket,
+                                             sock_mod = ejabberd_tls},
+	    case ejabberd_tls:recv_data(TLSSocket, <<"">>) of
+                {ok, TLSData} ->
+                    {noreply, process_data(TLSData, NewState), ?HIBERNATE_TIMEOUT};
+                {error, _Reason} ->
+                    {stop, normal, NewState}
+            end;
+        {error,_Reason} -> {stop, normal, State}
```

Comment:
at least metric and INFO_MSG is needed.
I know, the original code was not doing it too, but I had a client, who asked to add useful SSL metrics ><.

Reply:
it is "failed to establish connection" error, i think this one is logged, not indirectly, but still... if you need metrics - think about requirements first, and let's add it in another PR

Reply:
failed_tcp_to_tls is perfectly fine metric for me. Requirements:
- should be incremented each time tcp_to_tls fails
- should be a spiral type
- why? can help to find out, that some of our clients fail on this step of connecting.

Reply:
:+1: for an info or warning message (preferably the former). I don't think we have a guarantee that TLS library will log anything here and it might be useful for debugging.

New metric may be a TODO comment for now.

Reply:
actually i wasn't talking about TLS library. c2s process monitors receiver process, once receiver fails c2s ends up with an error message.

Reply:
With `{stop, normal, State}` there won't be any meaningful error reported.

Reply:
your propositions?

Reply:
Either stop with error reason or at least log a debug message with a reason here.